clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3025,19 +3025,34 @@ class SYCLKernelNameTypeVisitor
         S.getASTContext().getLangOpts().SYCLUnnamedLambda;
     const DeclContext *DeclCtx = Tag->getDeclContext();
     if (DeclCtx && !UnnamedLambdaEnabled) {
-      auto *NameSpace = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
-      if (NameSpace && NameSpace->isStdNamespace()) {
-        S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << KernelNameType;
-        S.Diag(KernelInvocationFuncLoc, diag::note_invalid_type_in_sycl_kernel)
-            << /* kernel name cannot be a type in the std namespace */ 2
-            << QualType(Tag->getTypeForDecl(), 0);
-        IsInvalid = true;
-        return;
-      }
-      if (!DeclCtx->isTranslationUnit() && !isa<NamespaceDecl>(DeclCtx)) {
-        const bool KernelNameIsMissing = Tag->getName().empty();
-        if (KernelNameIsMissing) {
+
+      while (!DeclCtx->isTranslationUnit()) {
+        auto *NSDecl = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
+        if (NSDecl && NSDecl->isStdNamespace()) {
```

Comment:
You're right that `std::test::U` should fail, however the loop makes me concerned that the visitor isn't doing what it is supposed to.  Why is it not doing a 'visit' of the namespace and catching it that way?  I would expect each of the decl-contexts to be visited, not need to be looped here.

Reply:
This is a `TypeVisitor` and `ConstTemplateArgumentVisitor`, we have to inherit it from some "declaration context visitor" to do so, if this thing exists at all.

Reply:
Ah, huh... I figured we would use something like the text-dumpers use, which visit each node along the way individually.  They seem to better fit the integration-header/kernel name checking.

Reply:
Do you mean things like `TextNodeDumper`? i.e. this one https://clang.llvm.org/doxygen/classclang_1_1TextNodeDumper.html .
The original idea was to make our sisitor similar to this one. That is why we used `TypeVisitor` and `ConstTemplateArgumentVisitor` as bases, as well as `TextNodeDumper` does. We could take a look on how it handles namespaces and other declaration contexts, though.

Reply:
Right, my initial thought was something like the `TextNodeDumper` or the `JSONNodeDumper` with the intent of avoiding all the looping we were doing over decl-context everywhere.

Reply:
> Right, my initial thought was something like the `TextNodeDumper` or the `JSONNodeDumper` with the intent of avoiding all the looping we were doing over decl-context everywhere.

When the diagnostics were originally refactored,  the existing code/logic was re-written using Visitor classes for which TypeVisitor and ConstTemplateArgumentVisitor were most appropriate.
The whole topic of checking declaration contexts came up only recently with the need to support nested name types, which is why [DeclVisitor](https://clang.llvm.org/doxygen/classclang_1_1ConstDeclVisitor.html) was not used before.

Reply:
Thanks for the context.  

We definitely need to prevent std::something::SomethingElse, the fact that we aren't is likely a regression (as I seem to remember it worked before the refactor?  Unfortunate there were no tests).

This patch does NOT seem to do anything with the integration header though, which would need to be taught how to forward-declare these globally visible names.  That said, I don't think this necessary part of this is actually possible with integration-headers (see my other comment).

At the moment, I'd suggest a separate patch to fix the std::something::something_else case (as well as the something::<anon_ns>::another::type case), and we wait for the nested-struct version until we can get @mkinsner and the SYCL spec clarifications of what we have to do.

if we have to support this nested struct case, I think we have to figure out an alternative to integration headers for runtime support.

Reply:
> as I seem to remember it worked before the refactor? Unfortunate there were no tests)

Refactor did not break anything. std::something::SomethingElse was not diagnosed before. 
Same with anonymous NS.

Reply:
Interesting.  That diffs from my memory, but still needs fixing.

Reply:
You can look at the history. I can fix it, just clarifying that it was something not previously supported and something I encountered while working on the original issue.

Reply:
I believe you :)  I could swear it was caught at SOME point (as I was sure I encountered it at one point...), but it doesn't mean it survived until the refactor in question I guess.